File: kits/data-integration/providers/structure_detector/KvDetector.swift
// Key-value detector — finds "Key: Value", "Key = Value", "Key -> Value" patterns
// Infers value types: numbers, dates, booleans, URLs, plain strings

import Foundation

private struct KvPattern {
    let regex: NSRegularExpression
    let confidence: Double
}

public final class KvDetectorProvider {

    public init() {}

    private func inferValueType(_ raw: String) -> (value: Any, type: String) {
        let trimmed = raw.trimmingCharacters(in: .whitespaces)
        let lower = trimmed.lowercased()

        // Boolean
        if ["true", "yes", "on"].contains(lower) { return (true, "boolean") }
        if ["false", "no", "off"].contains(lower) { return (false, "boolean") }

        // Integer
        if let n = Int(trimmed), trimmed.count <= 15 { return (n, "number") }

        // Float
        if trimmed.contains("."), let f = Double(trimmed) { return (f, "number") }

        // ISO date
        let datePattern = try? NSRegularExpression(pattern: #"^\d{4}-\d{2}-\d{2}(T\d{2}:\d{2}:\d{2})?"#)
        if let dp = datePattern, dp.firstMatch(in: trimmed, range: NSRange(trimmed.startIndex..., in: trimmed)) != nil {
            return (trimmed, "date")
        }

        // URL
        if trimmed.hasPrefix("http://") || trimmed.hasPrefix("https://") {
            return (trimmed, "url")
        }

        // Email
        let emailPattern = try? NSRegularExpression(pattern: #"^[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}$"#)
        if let ep = emailPattern, ep.firstMatch(in: trimmed, range: NSRange(trimmed.startIndex..., in: trimmed)) != nil {
            return (trimmed, "email")
        }

        return (trimmed, "string")
    }

    private func normalizeKey(_ raw: String) -> String {
        raw.trimmingCharacters(in: .whitespaces)
            .lowercased()
            .map { $0.isWhitespace ? "_" : $0 }
            .map(String.init)
            .joined()
            .filter { $0.isLetter || $0.isNumber || $0 == "_" }
    }

    private func buildPatterns() -> [KvPattern] {
        let specs: [(String, Double)] = [
            (#"^([A-Za-z][A-Za-z0-9 _\-]{0,50})\s*:\s+(.+)$"#, 0.92),
            (#"^([A-Za-z][A-Za-z0-9 _\-]{0,50})\s*=\s+(.+)$"#, 0.88),
            (#"^([A-Za-z][A-Za-z0-9 _\-]{0,50})\s*->\s+(.+)$"#, 0.85),
            (#"^([A-Za-z][A-Za-z0-9 _\-]{0,50})\s*=>\s+(.+)$"#, 0.85),
            (#"^([A-Za-z][A-Za-z0-9 _\-]{0,50})\s*\u{2192}\s+(.+)$"#, 0.85),
            (#"^([A-Za-z][A-Za-z0-9 _\-]{0,50})\s+-\s+(.+)$"#, 0.70),
        ]
        return specs.compactMap { (pattern, conf) in
            guard let re = try? NSRegularExpression(pattern: pattern, options: .anchorsMatchLines) else { return nil }
            return KvPattern(regex: re, confidence: conf)
        }
    }

    public func detect(content: Any, existingStructure: [String: Any], config: DetectorConfig) throws -> [Detection] {
        let text: String
        if let s = content as? String { text = s }
        else { text = String(describing: content) }

        let threshold = config.confidenceThreshold ?? 0.5
        let patterns = buildPatterns()
        var detections: [Detection] = []
        var seen = Set<String>()

        for line in text.components(separatedBy: .newlines) {
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            guard !trimmed.isEmpty else { continue }
            let lineRange = NSRange(trimmed.startIndex..., in: trimmed)

            for pattern in patterns {
                guard let match = pattern.regex.firstMatch(in: trimmed, range: lineRange) else { continue }
                guard match.numberOfRanges >= 3,
                      let keyRange = Range(match.range(at: 1), in: trimmed),
                      let valRange = Range(match.range(at: 2), in: trimmed) else { continue }

                let rawKey = String(trimmed[keyRange])
                let rawValue = String(trimmed[valRange])
                let key = normalizeKey(rawKey)

                guard !key.isEmpty, key.count <= 50 else { continue }
                guard !rawValue.trimmingCharacters(in: .whitespaces).isEmpty else { continue }
                guard !seen.contains(key) else { continue }
                seen.insert(key)

                let (value, vtype) = inferValueType(rawValue)
                var confidence = pattern.confidence
                if vtype != "string" { confidence = min(confidence + 0.05, 0.99) }
                if rawKey.trimmingCharacters(in: .whitespaces).count <= 2 { confidence -= 0.15 }
                guard confidence >= threshold else { continue }

                detections.append(Detection(
                    field: key, value: value, type: vtype,
                    confidence: confidence, evidence: trimmed
                ))
                break // first matching separator wins
            }
        }

        return detections
    }

    public func appliesTo(contentType: String) -> Bool {
        ["text/plain", "text/markdown", "text/yaml", "application/x-yaml"].contains(contentType)
    }
}

File: kits/data-integration/providers/structure_detector/kv_detector.rs
// Key-value detector — finds "Key: Value", "Key = Value", "Key -> Value" patterns
// Infers value types: numbers, dates, booleans, URLs, plain strings

use regex::Regex;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::{HashMap, HashSet};

pub const PROVIDER_ID: &str = "kv_detector";
pub const PLUGIN_TYPE: &str = "structure_detector";

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DetectorConfig {
    pub options: Option<HashMap<String, Value>>,
    pub confidence_threshold: Option<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Detection {
    pub field: String,
    pub value: Value,
    pub r#type: String,
    pub confidence: f64,
    pub evidence: String,
}

#[derive(Debug)]
pub enum DetectorError {
    ParseError(String),
    RegexError(String),
}

impl std::fmt::Display for DetectorError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            DetectorError::ParseError(msg) => write!(f, "Parse error: {}", msg),
            DetectorError::RegexError(msg) => write!(f, "Regex error: {}", msg),
        }
    }
}

fn infer_value_type(raw: &str) -> (Value, &'static str) {
    let trimmed = raw.trim();

    // Boolean
    let lower = trimmed.to_lowercase();
    if matches!(lower.as_str(), "true" | "yes" | "on") {
        return (Value::Bool(true), "boolean");
    }
    if matches!(lower.as_str(), "false" | "no" | "off") {
        return (Value::Bool(false), "boolean");
    }

    // Integer
    if let Ok(n) = trimmed.parse::<i64>() {
        if trimmed.len() <= 15 {
            return (Value::Number(n.into()), "number");
        }
    }

    // Float
    if trimmed.contains('.') {
        if let Ok(f) = trimmed.parse::<f64>() {
            if let Some(n) = serde_json::Number::from_f64(f) {
                return (Value::Number(n), "number");
            }
        }
    }

    // ISO date
    let date_re = Regex::new(r"^\d{4}-\d{2}-\d{2}(T\d{2}:\d{2}:\d{2})?").unwrap();
    if date_re.is_match(trimmed) {
        return (Value::String(trimmed.to_string()), "date");
    }

    // URL
    if trimmed.starts_with("http://") || trimmed.starts_with("https://") {
        return (Value::String(trimmed.to_string()), "url");
    }

    // Email
    let email_re = Regex::new(r"^[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}$").unwrap();
    if email_re.is_match(trimmed) {
        return (Value::String(trimmed.to_string()), "email");
    }

    (Value::String(trimmed.to_string()), "string")
}

fn normalize_key(raw: &str) -> String {
    raw.trim()
        .to_lowercase()
        .chars()
        .map(|c| if c.is_whitespace() { '_' } else { c })
        .filter(|c| c.is_alphanumeric() || *c == '_')
        .collect()
}

struct KvPattern {
    regex: Regex,
    confidence: f64,
}

pub struct KvDetectorProvider;

impl KvDetectorProvider {
    pub fn new() -> Self { Self }

    fn build_patterns() -> Vec<KvPattern> {
        vec![
            KvPattern {
                regex: Regex::new(r"^([A-Za-z][A-Za-z0-9 _\-]{0,50})\s*:\s+(.+)$").unwrap(),
                confidence: 0.92,
            },
            KvPattern {
                regex: Regex::new(r"^([A-Za-z][A-Za-z0-9 _\-]{0,50})\s*=\s+(.+)$").unwrap(),
                confidence: 0.88,
            },
            KvPattern {
                regex: Regex::new(r"^([A-Za-z][A-Za-z0-9 _\-]{0,50})\s*->\s+(.+)$").unwrap(),
                confidence: 0.85,
            },
            KvPattern {
                regex: Regex::new(r"^([A-Za-z][A-Za-z0-9 _\-]{0,50})\s*=>\s+(.+)$").unwrap(),
                confidence: 0.85,
            },
            KvPattern {
                regex: Regex::new(r"^([A-Za-z][A-Za-z0-9 _\-]{0,50})\s*\u{2192}\s+(.+)$").unwrap(),
                confidence: 0.85,
            },
            KvPattern {
                regex: Regex::new(r"^([A-Za-z][A-Za-z0-9 _\-]{0,50})\s+-\s+(.+)$").unwrap(),
                confidence: 0.70,
            },
        ]
    }

    pub fn detect(
        &self,
        content: &str,
        _existing: &HashMap<String, Value>,
        config: &DetectorConfig,
    ) -> Result<Vec<Detection>, DetectorError> {
        let threshold = config.confidence_threshold.unwrap_or(0.5);
        let patterns = Self::build_patterns();
        let mut detections = Vec::new();
        let mut seen = HashSet::new();

        for line in content.lines() {
            let trimmed = line.trim();
            if trimmed.is_empty() { continue; }

            for pattern in &patterns {
                if let Some(cap) = pattern.regex.captures(trimmed) {
                    let raw_key = &cap[1];
                    let raw_value = &cap[2];
                    let key = normalize_key(raw_key);

                    if key.is_empty() || key.len() > 50 { continue; }
                    if raw_value.trim().is_empty() { continue; }
                    if !seen.insert(key.clone()) { continue; }

                    let (value, vtype) = infer_value_type(raw_value);
                    let mut confidence = pattern.confidence;
                    if vtype != "string" {
                        confidence = (confidence + 0.05).min(0.99);
                    }
                    if raw_key.trim().len() <= 2 {
                        confidence -= 0.15;
                    }
                    if confidence < threshold { continue; }

                    detections.push(Detection {
                        field: key,
                        value,
                        r#type: vtype.to_string(),
                        confidence,
                        evidence: trimmed.to_string(),
                    });
                    break; // first matching separator wins
                }
            }
        }

        Ok(detections)
    }

    pub fn applies_to(&self, content_type: &str) -> bool {
        matches!(content_type, "text/plain" | "text/markdown" | "text/yaml" | "application/x-yaml")
    }
}
